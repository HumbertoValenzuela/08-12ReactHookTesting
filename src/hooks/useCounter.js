import { useState } from "react"


export const useCounter = ( initialState = 10) => {
  
    const [state, setState] = useState(initialState);

    const increment = () => {
        setState ( state + 1);
    };

    const decrement = () => {
        setState ( state - 1);
    };

    return {
        state, 
        increment,
        decrement
    };
}

// useCounter con argumento factor
export const useCounterArgumento = ( initialState = 100) => {
  
    const [state1, setState1] = useState(initialState);

    const increment1 = () => {
        setState1 ( state1 + 1);
    };

    const reset = () => {
        setState1(initialState);
    };

    const decrement1 = () => {
        setState1 ( state1 - 1);
    };

    return {
        state1, 
        reset,
        increment1,
        decrement1
    };
};

//Al llamar useCounterArgumento tenemos un return. Contiene un state1(numero) y 3 funciones