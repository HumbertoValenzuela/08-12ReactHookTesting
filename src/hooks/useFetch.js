// rafc
// 10._useFetch_-_CustomHook. Peticiones GET a cualquier URL
// https://breakingbadapi.com/
import { useEffect, useRef, useState } from "react";

export const useFetch = ( url ) => {

    // useRef por defecto cuando es llamado, el componente esta montado porque se esta renderizando la primera vez
    // La idea de ismounted es que mantenga la referencia cuando este hook esta vivo o cuando el componente que lo usa sigue montado
    // Cuando se cambia los valores de ismounted, no se quiere lanzar una renderizaci칩n del componente. Simplemente se mantiene la 
    // referencia al valor
    const isMounted = useRef(true);

    useEffect( () => {

        // Cuando se desmonte
        return () => {
            isMounted.current = false;
        }

    }, [])

    // estado inicial data: null, loading: true, error: null
    const [ state, setState ] = useState( { data: null, loading: true, error: null } );
    
    // Cambia el url realizar un Effect
    useEffect( () => {

        // Al comienzo de la petici칩n, modificar State para que muestre el loading
        setState( { data: null, loading: true, error: null } );

        fetch( url )
            .then( resp => resp.json() ) // retorna otra promesa
            .then( data => {
                
                // Establecer el primer cambio de State
                // setTimeout(() => {
                    // Para impedir el error usando useRef
                    // true esta montado, false no esta montado
                    if ( isMounted.current ) {
                        setState( { 
                            loading: false,
                            error: null,
                            data // ser치 igual a la respuesta del EndPoint
                         })                        
                    }  else {
                        console.log('setState no se llam칩');
                    }                  
                // }, 4000);
                
               
            })
            .catch( () => {
                setState({
                    data:null,
                    loading: false,
                    error: 'No se pudo cargar la info'
                })
            })
    }, [url] )

    return state;
}
