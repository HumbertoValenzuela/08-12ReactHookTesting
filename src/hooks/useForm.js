import { useState } from "react"

// Custom hook maneja el formulario
export const useForm = (initialState = {} ) => {

    const [valueForm, setValueForm] = useState(initialState);
    // Limpiar input
    const reset = () => {
        setValueForm( initialState );
    }
    // Copia lo anterior y agrega uno nuevo
    const handleInputChange = ( { target } ) => {

        setValueForm( { ...valueForm, [ target.name ]: target.value } 
        );
    }
    
    return [ valueForm, handleInputChange , reset];
}


