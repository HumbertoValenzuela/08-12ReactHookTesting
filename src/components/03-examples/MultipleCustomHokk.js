import React from 'react'
import { useCounterArgumento } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch'

export const MultipleCustomHokk = () => {

    // custom hook que maneja el counter
    const {state1, increment1} = useCounterArgumento(1);

    // componente useFetch
    const { loading, data } = useFetch( `https://www.breakingbadapi.com/api/quotes/${ state1 }`);
    
    // 11._useFetch_+_useCounter Mostrar en pantalla
    // quote_id: 3, quote: "IFT", author: "Skyler White", series: "Breaking Bad"

    //!!data && data[0] Si viene la data entonces extrae la data[0]
    // !! doble negaci√≥n transforma el null en un false
    // !null es true y !!null false
    // !null && 'hola' Si es true entonces hola
    // !!null && 'hola' Si no tiene valor entonces no va a ejecutar ni evaluar. retorna false
    const { author, quote} = !!data && data[0];
    // console.log(author, quote);
    return (
        <div>
            <h4>Breaking Bad Quotes</h4>          
            {
                loading
                ?
                    (
                        <div className="alert alert-info text-center">
                            Loading...
                        </div>
                    )
                :
                    (
                        <blockquote className="blockquote text-right">
                            <p className="mb-0">{ quote }</p>
                            <footer className="blockquote-footer">{ author }</footer>
                        </blockquote>
                    )
            }

            <button 
                className="btn btn-primary"
                onClick={increment1}
                >
                Siguiente quote
            </button>  
             
        </div>
    )
}
