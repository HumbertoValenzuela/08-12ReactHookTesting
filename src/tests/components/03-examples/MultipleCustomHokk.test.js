import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHokk } from '../../../components/03-examples/MultipleCustomHokk';
import { useFetch } from '../../../hooks/useFetch';
import { useCounterArgumento } from '../../../hooks/useCounter';

// jest.mock: cuando yo vaya a utilizar en este archivo, entonces en lugar de usar el useFetch, vamos a utlizar una implementación que me invento del useFetch
jest.mock( '../../../hooks/useFetch');
jest.mock( '../../../hooks/useCounter');
//  TypeError: Cannot destructure property 'loading' of '(0 , _useFetch.useFetch)(...)' as it is undefined.
// Es un error del componente no de la prueba



// Renderizar de manera condicional
describe('Pruebas en ', () => {

    useCounterArgumento.mockReturnValue( {
        state1: 1,
        increment1: () => {}
    });

    test('debe de mostrar snapshot por defecto ', () => {

        useFetch.mockReturnValue( {
            data:null,
            loading:true,
            error:null
        });

        const wrapper = shallow(<MultipleCustomHokk />);
        expect(wrapper).toMatchSnapshot();
    });

    // Mandar info de prueba hacia un Custom Hook
    test('debe de mostrar la info real', () => {
        // useFetch Interesa la info que retorna
        // un Mook, es decir, no interesa la implementación. Interesa los resultados
        useFetch.mockReturnValue( {
            data: [{
                author: 'Humberto',
                quote: 'Hola Mundo'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHokk />);
        // console.log( wrapper.html());
        // <div><h4>Breaking Bad Quotes</h4><blockquote class="blockquote text-right"><p class="mb-0">Hola Mundo</p><footer class="blockquote-footer">Humberto</footer></blockquote><button class="btn btn-primary">Siguiente quote</button></div>

        // Busquedas en el wrapper por una clase alert. No debería de existir si tengo información
        expect( wrapper.find('.alert').exists() ).toBe(false);
        // se espera la clase .mb-0 el texto sea Hola Mundo
        expect( wrapper.find('.mb-0').text().trim() ).toBe('Hola Mundo');
        expect( wrapper.find('footer').text().trim() ).toBe('Humberto');
    })
    
})
