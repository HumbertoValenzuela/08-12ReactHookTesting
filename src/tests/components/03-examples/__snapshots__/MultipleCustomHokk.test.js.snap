// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en  debe de mostrar snapshot por defecto  1`] = `
<div>
  <h4>
    Breaking Bad Quotes
  </h4>
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</div>
`;
